core/distributed/vector.cpp

```diff
@@ -0,0 +1,472 @@
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/distributed/vector.hpp>
+
+
+#include "core/distributed/vector_kernels.hpp"
+#include "core/matrix/dense_kernels.hpp"
+
+
+namespace gko {
+namespace distributed {
+namespace vector {
+namespace {
+
+
+GKO_REGISTER_OPERATION(compute_squared_norm2, dense::compute_squared_norm2);
+GKO_REGISTER_OPERATION(compute_sqrt, dense::compute_sqrt);
+GKO_REGISTER_OPERATION(outplace_absolute_dense, dense::outplace_absolute_dense);
+GKO_REGISTER_OPERATION(build_local, distributed_vector::build_local);
+
+
+}  // namespace
+}  // namespace vector
+
+
+dim<2> compute_global_size(mpi::communicator comm, dim<2> local_size)
+{
+    size_type num_global_rows = local_size[0];
+    comm.all_reduce(&num_global_rows, 1, MPI_SUM);
+    return {num_global_rows, local_size[1]};
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::apply_impl(const LinOp* b, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::apply_impl(const LinOp* alpha, const LinOp* b,
+                                   const LinOp* beta, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+template <typename ValueType>
+Vector<ValueType>::Vector(std::shared_ptr<const Executor> exec,
+                          mpi::communicator comm, dim<2> global_size,
+                          dim<2> local_size)
+    : Vector(exec, comm, global_size, local_size, local_size[1])
+{}
+
+
+template <typename ValueType>
+Vector<ValueType>::Vector(std::shared_ptr<const Executor> exec,
+                          mpi::communicator comm, dim<2> global_size,
+                          dim<2> local_size, size_type stride)
+    : EnableLinOp<Vector<ValueType>>{exec, global_size},
+      DistributedBase{comm},
+      local_{exec, local_size, stride}
+{
+    GKO_ASSERT_EQUAL_COLS(global_size, local_size);
+}
+
+template <typename ValueType>
+Vector<ValueType>::Vector(std::shared_ptr<const Executor> exec,
+                          mpi::communicator comm, dim<2> global_size,
+                          local_vector_type* local_vector)
+    : EnableLinOp<Vector<ValueType>>{exec, global_size},
+      DistributedBase{comm},
+      local_{exec}
+{
+    local_vector->move_to(&local_);
+}
+
+
+template <typename ValueType>
+Vector<ValueType>::Vector(std::shared_ptr<const Executor> exec,
+                          mpi::communicator comm,
+                          local_vector_type* local_vector)
+    : EnableLinOp<Vector<ValueType>>{exec, {}},
+      DistributedBase{comm},
+      local_{exec}
+{
+    this->set_size(compute_global_size(comm, local_vector->get_size()));
+    local_vector->move_to(&local_);
```

Comment:
yes, the idea is that the distributed vector takes ownership of the local vector. There is no need to have it twice, since you can get read/write access from the distributed vector.

Reply:
moving in this case means just swapping a few pointers, no actual data movement

Reply:
Yes, but original object can not be used anymore.
If user creates the dense view, then the copy only copies the view pointer, the data is not copied.

Reply:
yes, and that is intended (if I understand you correctly). That allows us to wrap user data also into distributed objects. A copy/clone of a non-composite object (like Dense) always performs a deep copy, i.e. it copies the values into a new array. If you want to reuse the storage, you need to move from it.